File: bioinformatics/multi_channel.r
# process multi-channel (mtch) data. Use global variable `channels`.


# function: consolidate result_list_by_channel (rlc) for one element in the one-channel result. 
consolidate_rlc_per_element <- function(result_ele_name, result_list_by_channel) {
    
    erfc <- result_list_by_channel[[1]][[result_ele_name]] # element of interest in the result for the first channel
    channels <- names(result_list_by_channel)
    
    if (class(erfc) == 'matrix') {
        consolidated_ele <- array(NA, 
                                  dim = c(length(channels), dim(erfc)[1], dim(erfc)[2]),
                                  dimnames = list(channels, rownames(erfc), colnames(erfc)))
        for (channel in channels) consolidated_ele[channel,,] <- result_list_by_channel[[channel]][[result_ele_name]] # lapply only make changes in the local scope.
    } else {
        consolidated_ele <- lapply(channels, function(channel) result_list_by_channel[[channel]][[result_ele_name]]) 
        names(consolidated_ele) <- channels }
    
    return(consolidated_ele)
    }


# function: run qPCR functions on multi-channel fluorescence data and output consolidated results. original matrix will become 3-D array, original string or list will be come list.
process_mtch <- function( 
                         iterable_by_channel, # must be a named list or vector, where names are channels, to be inherited by result_list_by_channel
                         func, 
                         ...) {
    
    result_list_by_channel <- lapply(iterable_by_channel, func, ...) # result_list_by_channel will inherit names from iterable_by_channel
    
    result_ele_names <- names(result_list_by_channel[[1]])
    
    result_consoli <- lapply(result_ele_names, consolidate_rlc_per_element, result_list_by_channel)
    names(result_consoli) <- result_ele_names
    
    return(list('pre_consoli'=result_list_by_channel,
                'post_consoli'=result_consoli))
    }

#

File: bioinformatics/QpcrAnalysis/multi_channel.jl
# multi-channel.jl
#
# process multi-channel (mtch) data. 
# these functions are not currently used

# Top-level function: run qPCR functions on multi-channel fluorescence data and output consolidated results. original matrix will become 3-D array, original string or list will be come list.
function process_mtch(
    input_dict_bych::OrderedDict, # keyed by channels
    arydims2to3::Bool,
    func::Function,
    func_args...
    )

    channels = get_ordered_keys(input_dict_bych)

    result_dict_bych = OrderedDict([
        channel => func(input_dict_bych[channel], func_args...)
        for channel in channels])

    result_keys = get_ordered_keys(result_dict_bych[channels[1]])

    result_consoli = OrderedDict([
        result_key => consolidate_rdc_per_element(result_key, result_dict_bych, arydims2to3)
        for result_key in result_keys])

    return OrderedDict(
        "pre_consoli" => result_dict_bych,
        "post_consoli" => result_consoli)

end


# functions called by `process_mtch`

# function: consolidate for one result element from organization of channel then result element (as in `result_dict_bych`) into organization of result element then channel
function consolidate_rdc_per_element(
    result_key,
    result_dict_bych::OrderedDict,
    arydims2to3::Bool
    )

    channels = get_ordered_keys(result_dict_bych)
    erfc = result_dict_bych[channels[1]][result_key] # element of interest in the result for the first channel

    if arydims2to3 && isa(erfc, AbstractArray) && ndims(erfc) == 2
        consolidated_ele = cat(3, map(channels) do channel
            result_dict_bych[channel][result_key]
        end...) # The resulting DataArray is indexed by the linear indice of channel in `channels` instead of the actual channel values.
    else
        consolidated_ele = OrderedDict(map(channels) do channel
            channel => result_dict_bych[channel][result_key]
        end)
    end

    return consolidated_ele

end


#
